From: Daniel van Vugt <daniel.van.vugt@canonical.com>
Date: Thu, 3 Aug 2023 16:11:59 +0800
Subject: backends/x11: Trap errors from XIAllowTouchEvents

And report them as debug messages instead of crashing. We don't want them
to be visible usually because failures are expected in the autodeny
code path.

Bug: https://gitlab.gnome.org/GNOME/mutter/-/issues/2808
Bug-Ubuntu: https://launchpad.net/bugs/2029413
Forwarded: https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/3146
Applied-upstream: 44.5, commit:7fc14155
---
 src/backends/x11/meta-backend-x11.c | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/src/backends/x11/meta-backend-x11.c b/src/backends/x11/meta-backend-x11.c
index 0f7fe08..e647fc8 100644
--- a/src/backends/x11/meta-backend-x11.c
+++ b/src/backends/x11/meta-backend-x11.c
@@ -700,6 +700,7 @@ meta_backend_x11_finish_touch_sequence (MetaBackend          *backend,
   MetaBackendX11 *x11 = META_BACKEND_X11 (backend);
   MetaBackendX11Private *priv = meta_backend_x11_get_instance_private (x11);
   int event_mode;
+  int err;
 
   if (state == META_SEQUENCE_ACCEPTED)
     event_mode = XIAcceptTouch;
@@ -708,10 +709,21 @@ meta_backend_x11_finish_touch_sequence (MetaBackend          *backend,
   else
     g_return_if_reached ();
 
+  meta_clutter_x11_trap_x_errors ();
   XIAllowTouchEvents (priv->xdisplay,
                       META_VIRTUAL_CORE_POINTER_ID,
                       clutter_event_sequence_get_slot (sequence),
                       DefaultRootWindow (priv->xdisplay), event_mode);
+  /* We need to XSync so meta_clutter_x11_untrap_x_errors doesn't miss errors
+   * from XIAllowTouchEvents.
+   */
+  XSync (priv->xdisplay, False);
+  err = meta_clutter_x11_untrap_x_errors ();
+  if (err)
+    {
+      g_debug ("XIAllowTouchEvents failed event_mode %d with error %d",
+               event_mode, err);
+    }
 
   if (state == META_SEQUENCE_REJECTED)
     {
